include/wlr/types/wlr_output_layer.h

```diff
@@ -0,0 +1,83 @@
+/*
+ * This an unstable interface of wlroots. No guarantees are made regarding the
+ * future consistency of this API.
+ */
+#ifndef WLR_USE_UNSTABLE
+#error "Add -DWLR_USE_UNSTABLE to enable unstable wlroots features"
+#endif
+
+#ifndef WLR_TYPES_WLR_OUTPUT_LAYER_H
+#define WLR_TYPES_WLR_OUTPUT_LAYER_H
+
+/**
+ * Output layers provide an API to display buffers without rendering them. It
+ * relies on backend features which are not always available: the backend can
+ * refuse to display a layer. However when the backend accepts to display a
+ * layer, performance and battery usage will be improved.
+ *
+ * Operations on output layers are double-buffered. An output commit is needed
+ * to apply the pending state.
+ */
+
+#include <wayland-server-core.h>
+
+enum wlr_output_layer_state_field {
+	WLR_OUTPUT_LAYER_STATE_LINK = 1 << 0,
+	WLR_OUTPUT_LAYER_STATE_BUFFER = 1 << 1,
+	WLR_OUTPUT_LAYER_STATE_POSITION = 1 << 2,
+};
+
+struct wlr_output_layer_state {
+	uint32_t committed; // enum wlr_output_layer_state_field
+	struct wl_list link;
+	struct wlr_buffer *buffer; // only valid if WLR_OUTPUT_LAYER_STATE_BUFFER
+	int x, y; // only valid if WLR_OUTPUT_LAYER_STATE_POSITION
+};
+
+struct wlr_output_layer {
+	struct wlr_output *output;
+	/** If true, the backend has accepted to display the layer. If false, the
+	 * compositor needs to manually render the layer. After each output commit,
```

Comment:
- Should manual rendering be through a large layer that was initialized earlier, or through a buffer attached directly on the wlr_output?
- How do layers interact with buffers attached directly to a wlr_output? E.g. ordering - place above/below only operate on layers after all.

Reply:
That's a good question.

Right now, the output's primary buffer (set via `wlr_output_attach_render` or `wlr_output_attach_buffer`) is always used, cannot be NULL, and is assumed to be the fallback in case it's not possible to display all output layers (ie, the compositor will paint layers that can't be displayed onto the primary buffer). The primary buffer is always below all output layers.

We should at least document all of that. We should also wonder whether this is a good design.

I think long-term we'll want to have the primary buffer be a regular layer just like the other layers. The main motivation is that both KMS and Wayland allow the primary buffer to be placed below layers and to set the same scaling/cropping/etc properties on both primary and layers.

However, we probably want to keep the old `wlr_output_attach_render`/`wlr_output_attach_buffer` functions working, at least for some time. If backends can take the primary buffer in a regular layer, these functions can probably just internally create a layer.

So I _think_ it's fine to keep the status quo in this PR as a first step, then upgrade to the long-term plan described above.

Reply:
> Right now, the output's primary buffer (set via `wlr_output_attach_render` or `wlr_output_attach_buffer`) is always used, cannot be NULL, and is assumed to be the fallback in case it's not possible to display all output layers (ie, the compositor will paint layers that can't be displayed onto the primary buffer). The primary buffer is always below all output layers.

Hmm, what if the declined layer has a z-ordering above other opaque layers, so that it cannot be painted from below? We'd have to bail all intersecting layers as well in that case, or carve a transparent section into them.

Reply:
That is correct. libliftoff should handle this case already. There are quite a few test cases, this one should cover the case you describe:

https://github.com/emersion/libliftoff/blob/799dddd6c449d7b34bedb24e413b4d754f05a9ab/test/test_alloc.c#L141

Reply:
Ah, so the idea is that the implementation of wlr_output_layer guarantees that intersecting layers are also declined? I was worried the compositor would need to handle this. Such guarantee would be nice to mention here.

I'm fine with having the regular output buffer act as composition buffer, but that would also be nice to have documented as it not obvious.

Reply:
Added some documentation about this.